Mathlib/Data/Matroid/Loop.lean

```diff
@@ -401,4 +427,298 @@ lemma closure_inter_setOf_isNonloop_eq (M : Matroid α) (X : Set α) :
 
 end IsNonloop
 
+section IsColoop
+
+variable {B K : Set α}
+
+/-- A coloop is a loop of the dual matroid.
+See `Matroid.isColoop_tfae` for a number of equivalent definitions. -/
+def IsColoop (M : Matroid α) (e : α) : Prop := M✶.IsLoop e
+
+/-- `M.coloops` is the set of coloops of `M`. -/
+def coloops (M : Matroid α) := M✶.loops
+
+@[aesop unsafe 20% (rule_sets := [Matroid])]
+lemma IsColoop.mem_ground (he : M.IsColoop e) : e ∈ M.E :=
+  @IsLoop.mem_ground α (M✶) e he
+
+@[aesop unsafe 20% (rule_sets := [Matroid])]
+lemma coloops_subset_ground (M : Matroid α) : M.coloops ⊆ M.E :=
+  fun _ ↦ IsColoop.mem_ground
+
+lemma isColoop_iff_mem_loops : M.IsColoop e ↔ e ∈ M.coloops := Iff.rfl
+
+@[simp]
+lemma dual_loops : M✶.loops = M.coloops := rfl
+
+@[simp]
+lemma dual_coloops : M✶.coloops = M.loops := by
+  rw [coloops, dual_dual]
+
+lemma IsColoop.dual_isLoop (he : M.IsColoop e) : M✶.IsLoop e :=
+  he
+
+lemma IsColoop.isCocircuit (he : M.IsColoop e) : M.IsCocircuit {e} :=
+  IsLoop.isCircuit he
+
+lemma IsLoop.dual_isColoop (he : M.IsLoop e) : M✶.IsColoop e :=
+  by rwa [IsColoop, dual_dual]
+
+@[simp]
+lemma dual_isColoop_iff_isLoop : M✶.IsColoop e ↔ M.IsLoop e :=
+  ⟨fun h ↦ by rw [← dual_dual M]; exact h.dual_isLoop, IsLoop.dual_isColoop⟩
+
+@[simp]
+lemma dual_isLoop_iff_isColoop : M✶.IsLoop e ↔ M.IsColoop e :=
+  ⟨fun h ↦ by rw [← dual_dual M]; exact h.dual_isColoop, IsColoop.dual_isLoop⟩
+
+lemma singleton_isCocircuit : M.IsCocircuit {e} ↔ M.IsColoop e := by
+  simp
+
+lemma isColoop_tfae (M : Matroid α) (e : α) : List.TFAE [
+    M.IsColoop e,
+    e ∈ M.coloops,
+    M.IsCocircuit {e},
+    ∀ ⦃B⦄, M.IsBase B → e ∈ B,
+    (∀ ⦃C⦄, M.IsCircuit C → e ∉ C) ∧ e ∈ M.E,
+    (∀ X, e ∈ M.closure X ↔ e ∈ X) ∧ e ∈ M.E,
+    ¬ M.Spanning (M.E \ {e}) ] := by
```

Comment:
Is it possible to golf this proof using duality?

Reply:
I've pretty much golfed it as much as I can - duality is immediately useful for `1 <-> 4`, but the interaction between duality, closure and circuits is somewhat underdeveloped at this point in the import hierarchy.

Reply:
Ok, understood. Thanks for clarifying.

Reply:
I did just find a small improvement for one of the equivalences.